### PR TITLE
Add status tracking with loading spinners, clean up logging

### PR DESCRIPTION
**Pull request by BenTheElder**
- Adds a `--loglevel` flag for setting the logrus log level
- moves the default log level to warn
- adds a nice loading spinner system that tracks phases of creating the cluster for friendlier output while waiting for the cluster to boot

TODO (followup): 
- log kubeadm output at trace level (?) after upgrading logrus, or log it to a debug log file
- give helpful output when kubeadm init fails
- add preflight checks for docker, previous clusters, etc. #39 
- possibly just replace node.Run with RunQ

### CODE REVIEW

**Comment by BenTheElder** [comment]
cc @Katharine @munnerz 

This gives shiny output like:

```console
$ kind create cluster --name=asdf
Creating cluster 'kind-asdf' ...
 ✓ Ensuring node image (kindest/node:v1.11.3@sha256:855562266d5f647b5770a91c76a0579ed6487eb5bf1dfe908705dad285525483) 🖼
 ✓ [kind-asdf-control-plane] Creating node container 📦
 ✓ [kind-asdf-control-plane] Fixing mounts 🗻
 ✓ [kind-asdf-control-plane] Starting systemd 🖥
 ✓ [kind-asdf-control-plane] Waiting for docker to be ready 🐋
 ✓ [kind-asdf-control-plane] Starting Kubernetes (this may take a minute) ☸
Cluster creation complete. You can now use the cluster with:

export KUBECONFIG="/Users/bentheelder/.kube/kind-config-asdf"
kubectl cluster-info
```

**Comment by BenTheElder** [comment]
trying to record this with asciinema is amusing https://asciinema.org/a/kNsHMXcXpRAj4z0lUqM8l4CRV

**Comment by BenTheElder** [comment]
2x speed gif: ezgif-1-a66c6f2cdbf8

**Comment by neolit123** [comment]
nice :+1:

File: cmd/kind/kind.go [hunk 1]
```diff
@@ -20,40 +20,76 @@ package kind
...
 		Short: "kind is a tool for managing local Kubernetes clusters",
 		Long:  "kind creates and manages local Kubernetes clusters using Docker container 'nodes'",
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			return runE(flags, cmd, args)
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return cmd.Help()
+		},
 	}
+	cmd.PersistentFlags().StringVar(&flags.LogLevel, "loglevel", defaultLevel.String(), "logrus log level")
```
Comment (munnerz) [inline cmd/kind/kind.go]: Can we add `e.g. info, error, warn` or something to this? It gets quite frustrating having to remember what strings to use for all different log levels 😅

File: cmd/kind/kind.go [hunk 2]
```diff
@@ -20,40 +20,76 @@ package kind
...
 	cmd.AddCommand(create.NewCommand())
 	cmd.AddCommand(delete.NewCommand())
 	return cmd
 }
 
+func runE(flags *Flags, cmd *cobra.Command, args []string) error {
+	level := defaultLevel
+	parsed, err := log.ParseLevel(flags.LogLevel)
+	if err != nil {
+		log.Warnf("Invalid log level '%s', defaulting to '%s'", flags.LogLevel, level)
```
Comment (munnerz) [inline cmd/kind/kind.go]: nit: May be worth doing {something} to ensure this message is always visible? `defaultLevel` is currently `warn`, but I'm slightly concerned if this gets changed in future 🙄

File: pkg/cluster/node.go [hunk 3]
```diff
@@ -209,7 +238,7 @@ func tryUntil(until time.Time, try func() bool) bool {
 // LoadImages loads image tarballs stored on the node into docker on the node
 func (nh *nodeHandle) LoadImages() {
 	// load images cached on the node into docker
-	if err := nh.Run(
+	if err := nh.RunQ(
```
Comment (munnerz) [inline pkg/cluster/node.go]: It'd be awesome if we can make `--loglevel=debug` also print all of these messages, for debug purposes.

From what I can tell, with this patch there will no longer be any way to debug failures.

Also, if a particular step fails can we make it print out the stdout/stderr of that failed command?

Comment (BenTheElder) [reply pkg/cluster/node.go]: I was thinking maybe `--loglevel=trace` should do this after we upgrade logrus, but debug actually sounds good.

Definitely we should print out failed command output. Especially for kubeadm.

File: cmd/kind/kind.go (hunk 2 above)
Comment (BenTheElder) [reply cmd/kind/kind.go]: Hmm, we could just make this yet another thing that does not actually use logging, I almost wound up writing our own small logging package :-)

File: cmd/kind/kind.go (hunk 1 above)
Comment (BenTheElder) [reply cmd/kind/kind.go]: Yeah, this flag info is not particularly helpful.

Comment (BenTheElder) [reply cmd/kind/kind.go]: added this

File: pkg/cluster/node.go (hunk 3 above)
Comment (BenTheElder) [reply pkg/cluster/node.go]: I think what we really want is:
- when critical things fail, inform the user and tell them what they can do about it
- if the log level is cranked high enough, log even "silent" command output on exit

File: cmd/kind/kind.go (hunk 2 above)
Comment (BenTheElder) [reply cmd/kind/kind.go]: we're discussing broader logging options out of band, but this seems to be fine for now.

**Comment by BenTheElder** [comment]
Lots of log output will still mostly work with this, you can set `--loglevel` across all commands now.
FWIW, I've done quite a bit of `kind` debugging now, and the output from the command we call has not been helpful. I'll be PRing a kind command for obtaining useful debugging info in a follow up. https://github.com/kubernetes-sigs/kind/pull/69#discussion_r227423197

**Pull request merged**